File: open_source_projects/Deepbox-main/benchmarks/deepbox/03-linalg.ts
/**
 * Benchmark 03 — Linear Algebra
 * Deepbox vs NumPy/SciPy
 */

import {
	cholesky,
	cond,
	det,
	eigvalsh,
	inv,
	lstsq,
	lu,
	matrixRank,
	norm,
	pinv,
	qr,
	slogdet,
	solve,
	solveTriangular,
	svd,
	trace,
} from "deepbox/linalg";
import { dot, tensor } from "deepbox/ndarray";
import { createSuite, footer, header, run } from "../utils";

const suite = createSuite("linalg");
header("Benchmark 03 — Linear Algebra");

// ── Matrix generators ───────────────────────────────────

function seededRng(seed: number) {
	let s = seed >>> 0;
	return () => {
		s = (s * 1664525 + 1013904223) >>> 0;
		return (s / 2 ** 32) * 2 - 1;
	};
}

function randMatrix(rows: number, cols: number, seed: number) {
	const rand = seededRng(seed);
	const data: number[][] = [];
	for (let i = 0; i < rows; i++) {
		const row: number[] = [];
		for (let j = 0; j < cols; j++) row.push(rand());
		data.push(row);
	}
	return tensor(data);
}

function symPosDefMatrix(n: number, seed: number) {
	const rand = seededRng(seed);
	const A: number[][] = [];
	for (let i = 0; i < n; i++) {
		const row: number[] = [];
		for (let j = 0; j < n; j++) row.push(rand());
		A.push(row);
	}
	const result: number[][] = [];
	for (let i = 0; i < n; i++) {
		const row: number[] = [];
		for (let j = 0; j < n; j++) {
			let s = 0;
			for (let k = 0; k < n; k++) s += A[i][k] * A[j][k];
			row.push(s + (i === j ? n : 0));
		}
		result.push(row);
	}
	return tensor(result);
}

function upperTriangular(n: number, seed: number) {
	const rand = seededRng(seed);
	const data: number[][] = [];
	for (let i = 0; i < n; i++) {
		const row: number[] = [];
		for (let j = 0; j < n; j++) row.push(j >= i ? rand() + (i === j ? 1 : 0) : 0);
		data.push(row);
	}
	return tensor(data);
}

// Matrices
const m10 = randMatrix(10, 10, 42);
const m25 = randMatrix(25, 25, 42);
const m50 = randMatrix(50, 50, 42);
const m100 = randMatrix(100, 100, 42);
const m200 = randMatrix(200, 200, 42);
const sym25 = symPosDefMatrix(25, 42);
const sym50 = symPosDefMatrix(50, 42);
const sym100 = symPosDefMatrix(100, 42);
const _sym200 = symPosDefMatrix(200, 42);
const tri25 = upperTriangular(25, 42);
const tri50 = upperTriangular(50, 42);
const tri100 = upperTriangular(100, 42);
const b25 = randMatrix(25, 1, 99);
const b50 = randMatrix(50, 1, 99);
const b100 = randMatrix(100, 1, 99);
const _b200 = randMatrix(200, 1, 99);
const rect50x30 = randMatrix(50, 30, 42);
const rect100x50 = randMatrix(100, 50, 42);

// ── Determinant ─────────────────────────────────────────

run(suite, "det", "10x10", () => det(m10));
run(suite, "det", "25x25", () => det(m25));
run(suite, "det", "50x50", () => det(m50));
run(suite, "det", "100x100", () => det(m100));

// ── Trace ───────────────────────────────────────────────

run(suite, "trace", "25x25", () => trace(m25));
run(suite, "trace", "100x100", () => trace(m100));
run(suite, "trace", "200x200", () => trace(m200));

// ── Norm ────────────────────────────────────────────────

run(suite, "norm (fro)", "25x25", () => norm(m25));
run(suite, "norm (fro)", "50x50", () => norm(m50));
run(suite, "norm (fro)", "100x100", () => norm(m100));

// ── Condition Number ────────────────────────────────────

run(suite, "cond", "25x25", () => cond(m25));
run(suite, "cond", "50x50", () => cond(m50));

// ── Matrix Rank ─────────────────────────────────────────

run(suite, "matrixRank", "25x25", () => matrixRank(m25));
run(suite, "matrixRank", "50x50", () => matrixRank(m50));

// ── Slogdet ─────────────────────────────────────────────

run(suite, "slogdet", "25x25", () => slogdet(m25));
run(suite, "slogdet", "50x50", () => slogdet(m50));
run(suite, "slogdet", "100x100", () => slogdet(m100));

// ── Inverse ─────────────────────────────────────────────

run(suite, "inv", "10x10", () => inv(m10));
run(suite, "inv", "25x25", () => inv(m25));
run(suite, "inv", "50x50", () => inv(m50));
run(suite, "inv", "100x100", () => inv(m100));

// ── Pseudo-Inverse ──────────────────────────────────────

run(suite, "pinv", "25x25", () => pinv(m25));
run(suite, "pinv", "50x50", () => pinv(m50));
run(suite, "pinv", "50x30", () => pinv(rect50x30));

// ── SVD ─────────────────────────────────────────────────

run(suite, "svd", "10x10", () => svd(m10));
run(suite, "svd", "25x25", () => svd(m25));
run(suite, "svd", "50x50", () => svd(m50));
run(suite, "svd", "100x100", () => svd(m100), { iterations: 10 });

// ── QR ──────────────────────────────────────────────────

run(suite, "qr", "10x10", () => qr(m10));
run(suite, "qr", "25x25", () => qr(m25));
run(suite, "qr", "50x50", () => qr(m50));
run(suite, "qr", "100x100", () => qr(m100), { iterations: 10 });

// ── LU ──────────────────────────────────────────────────

run(suite, "lu", "10x10", () => lu(m10));
run(suite, "lu", "25x25", () => lu(m25));
run(suite, "lu", "50x50", () => lu(m50));
run(suite, "lu", "100x100", () => lu(m100));

// ── Cholesky ────────────────────────────────────────────

run(suite, "cholesky", "25x25", () => cholesky(sym25));
run(suite, "cholesky", "50x50", () => cholesky(sym50));
run(suite, "cholesky", "100x100", () => cholesky(sym100));

// ── Eigenvalues (symmetric) ─────────────────────────────

run(suite, "eigvalsh", "25x25", () => eigvalsh(sym25));
run(suite, "eigvalsh", "50x50", () => eigvalsh(sym50));
run(suite, "eigvalsh", "100x100", () => eigvalsh(sym100));

// ── Solve ───────────────────────────────────────────────

run(suite, "solve", "25x25", () => solve(sym25, b25));
run(suite, "solve", "50x50", () => solve(sym50, b50));
run(suite, "solve", "100x100", () => solve(sym100, b100));

// ── Solve Triangular ────────────────────────────────────

run(suite, "solveTriangular", "25x25", () => solveTriangular(tri25, b25));
run(suite, "solveTriangular", "50x50", () => solveTriangular(tri50, b50));
run(suite, "solveTriangular", "100x100", () => solveTriangular(tri100, b100));

// ── Least Squares ───────────────────────────────────────

run(suite, "lstsq", "50x30", () => lstsq(rect50x30, b50));
run(suite, "lstsq", "100x50", () => lstsq(rect100x50, b100));

// ── Matmul (dot) ────────────────────────────────────────

run(suite, "matmul", "25x25", () => dot(m25, m25));
run(suite, "matmul", "50x50", () => dot(m50, m50));
run(suite, "matmul", "100x100", () => dot(m100, m100));
run(suite, "matmul", "200x200", () => dot(m200, m200), { iterations: 10 });

footer(suite, "deepbox-linalg.json");

File: open_source_projects/Deepbox-main/benchmarks/python/03_linalg.py
"""
Benchmark 03 — Linear Algebra
NumPy / SciPy
"""

import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

import numpy as np
import scipy.linalg as la
from utils import run, create_suite, header, footer

suite = create_suite("linalg", "NumPy/SciPy")
header("Benchmark 03 — Linear Algebra", "NumPy/SciPy")

# ── Matrix generators ───────────────────────────────────

rng = np.random.RandomState(42)

def rand_matrix(n, m=None, seed=42):
    r = np.random.RandomState(seed)
    return r.randn(n, m if m else n)

def sym_pos_def(n, seed=42):
    r = np.random.RandomState(seed)
    A = r.randn(n, n)
    return A @ A.T + n * np.eye(n)

def upper_tri(n, seed=42):
    r = np.random.RandomState(seed)
    U = np.triu(r.randn(n, n))
    np.fill_diagonal(U, np.abs(np.diag(U)) + 1)
    return U

m10 = rand_matrix(10)
m25 = rand_matrix(25)
m50 = rand_matrix(50)
m100 = rand_matrix(100)
m200 = rand_matrix(200)
sym25 = sym_pos_def(25)
sym50 = sym_pos_def(50)
sym100 = sym_pos_def(100)
sym200 = sym_pos_def(200)
tri25 = upper_tri(25)
tri50 = upper_tri(50)
tri100 = upper_tri(100)
b25 = rand_matrix(25, 1, 99)
b50 = rand_matrix(50, 1, 99)
b100 = rand_matrix(100, 1, 99)
b200 = rand_matrix(200, 1, 99)
rect50x30 = rand_matrix(50, 30)
rect100x50 = rand_matrix(100, 50)

# ── Determinant ─────────────────────────────────────────

run(suite, "det", "10x10", lambda: np.linalg.det(m10))
run(suite, "det", "25x25", lambda: np.linalg.det(m25))
run(suite, "det", "50x50", lambda: np.linalg.det(m50))
run(suite, "det", "100x100", lambda: np.linalg.det(m100))

# ── Trace ───────────────────────────────────────────────

run(suite, "trace", "25x25", lambda: np.trace(m25))
run(suite, "trace", "100x100", lambda: np.trace(m100))
run(suite, "trace", "200x200", lambda: np.trace(m200))

# ── Norm ────────────────────────────────────────────────

run(suite, "norm (fro)", "25x25", lambda: np.linalg.norm(m25, 'fro'))
run(suite, "norm (fro)", "50x50", lambda: np.linalg.norm(m50, 'fro'))
run(suite, "norm (fro)", "100x100", lambda: np.linalg.norm(m100, 'fro'))

# ── Condition Number ────────────────────────────────────

run(suite, "cond", "25x25", lambda: np.linalg.cond(m25))
run(suite, "cond", "50x50", lambda: np.linalg.cond(m50))

# ── Matrix Rank ─────────────────────────────────────────

run(suite, "matrixRank", "25x25", lambda: np.linalg.matrix_rank(m25))
run(suite, "matrixRank", "50x50", lambda: np.linalg.matrix_rank(m50))

# ── Slogdet ─────────────────────────────────────────────

run(suite, "slogdet", "25x25", lambda: np.linalg.slogdet(m25))
run(suite, "slogdet", "50x50", lambda: np.linalg.slogdet(m50))
run(suite, "slogdet", "100x100", lambda: np.linalg.slogdet(m100))

# ── Inverse ─────────────────────────────────────────────

run(suite, "inv", "10x10", lambda: np.linalg.inv(m10))
run(suite, "inv", "25x25", lambda: np.linalg.inv(m25))
run(suite, "inv", "50x50", lambda: np.linalg.inv(m50))
run(suite, "inv", "100x100", lambda: np.linalg.inv(m100))

# ── Pseudo-Inverse ──────────────────────────────────────

run(suite, "pinv", "25x25", lambda: np.linalg.pinv(m25))
run(suite, "pinv", "50x50", lambda: np.linalg.pinv(m50))
run(suite, "pinv", "50x30", lambda: np.linalg.pinv(rect50x30))

# ── SVD ─────────────────────────────────────────────────

run(suite, "svd", "10x10", lambda: np.linalg.svd(m10))
run(suite, "svd", "25x25", lambda: np.linalg.svd(m25))
run(suite, "svd", "50x50", lambda: np.linalg.svd(m50))
run(suite, "svd", "100x100", lambda: np.linalg.svd(m100), iterations=10)

# ── QR ──────────────────────────────────────────────────

run(suite, "qr", "10x10", lambda: np.linalg.qr(m10))
run(suite, "qr", "25x25", lambda: np.linalg.qr(m25))
run(suite, "qr", "50x50", lambda: np.linalg.qr(m50))
run(suite, "qr", "100x100", lambda: np.linalg.qr(m100), iterations=10)

# ── LU ──────────────────────────────────────────────────

run(suite, "lu", "10x10", lambda: la.lu(m10))
run(suite, "lu", "25x25", lambda: la.lu(m25))
run(suite, "lu", "50x50", lambda: la.lu(m50))
run(suite, "lu", "100x100", lambda: la.lu(m100))

# ── Cholesky ────────────────────────────────────────────

run(suite, "cholesky", "25x25", lambda: np.linalg.cholesky(sym25))
run(suite, "cholesky", "50x50", lambda: np.linalg.cholesky(sym50))
run(suite, "cholesky", "100x100", lambda: np.linalg.cholesky(sym100))

# ── Eigenvalues (symmetric) ─────────────────────────────

run(suite, "eigvalsh", "25x25", lambda: np.linalg.eigvalsh(sym25))
run(suite, "eigvalsh", "50x50", lambda: np.linalg.eigvalsh(sym50))
run(suite, "eigvalsh", "100x100", lambda: np.linalg.eigvalsh(sym100))

# ── Solve ───────────────────────────────────────────────

run(suite, "solve", "25x25", lambda: np.linalg.solve(sym25, b25))
run(suite, "solve", "50x50", lambda: np.linalg.solve(sym50, b50))
run(suite, "solve", "100x100", lambda: np.linalg.solve(sym100, b100))

# ── Solve Triangular ────────────────────────────────────

run(suite, "solveTriangular", "25x25", lambda: la.solve_triangular(tri25, b25))
run(suite, "solveTriangular", "50x50", lambda: la.solve_triangular(tri50, b50))
run(suite, "solveTriangular", "100x100", lambda: la.solve_triangular(tri100, b100))

# ── Least Squares ───────────────────────────────────────

run(suite, "lstsq", "50x30", lambda: np.linalg.lstsq(rect50x30, b50, rcond=None))
run(suite, "lstsq", "100x50", lambda: np.linalg.lstsq(rect100x50, b100, rcond=None))

# ── Matmul ──────────────────────────────────────────────

run(suite, "matmul", "25x25", lambda: m25 @ m25)
run(suite, "matmul", "50x50", lambda: m50 @ m50)
run(suite, "matmul", "100x100", lambda: m100 @ m100)
run(suite, "matmul", "200x200", lambda: m200 @ m200, iterations=10)

footer(suite, "numpy-linalg.json")
